### PR TITLE
Split the non-linear solver

### PR DESCRIPTION
**Pull request by ajreynol**
This splits the "non-linear solver" from "NonlinearExtension". The non-linear solver is the module that implements the inference schemas whereas NonlinearExtension is the glue code that manages the solver(s) for non-linear.

This also involves moving utilities from the non-linear solver to their own file.

### CODE REVIEW

File: src/theory/arith/nl_constraint.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+   * this database.
+   */
+  void registerConstraint(Node atom);
+  /** get constraints
+   *
+   * Returns a map m such that whenever
+   * m[lit][x] = ( r, coeff, k ), then
+   * ( lit <=>  (coeff * x) <k> r )
+   */
+  std::map<Node, std::map<Node, ConstraintInfo> >& getConstraints();
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_constraint.h:63]: so `d_c_info` (the returned reference) is modifiable

File: src/theory/arith/nl_monomial.h [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+   * v1, ..., vn where |m(v1)| <= ... <= |m(vn)| after this method is invoked.
+   */
+  void sortVariablesByModel(std::vector<Node>& ms, NlModel& m);
+  /** Get monomial contains children map
+   *
+   * This maps monomials to other monomials that are contained in them, e.g.
+   * x^2 * y may map to { x, x^2, y } if these three terms exist have been
+   * registered to this class.
+   */
+  std::map<Node, std::vector<Node> >& getContainsChildrenMap();
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_monomial.h:103]: is it mutable?

File: src/theory/arith/nl_monomial.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  /** Get monomial contains parent map, reverse of above */
+  std::map<Node, std::vector<Node> >& getContainsParentMap();
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_monomial.h:105]: ditto

File: src/theory/arith/nl_solver.cpp [hunk 4]
```diff
@@ -0,0 +1,1583 @@
...
+      }
+      else
+      {
+        worklist.insert(worklist.end(), current.begin(), current.end());
+      }
+    }
+  }
+  return false;
+}
+NlSolver::NlSolver(TheoryArith& containing, NlModel& model)
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_solver.cpp:64]: please add an empty line

File: src/theory/arith/nl_solver.cpp [hunk 5]
```diff
@@ -0,0 +1,1583 @@
...
+                             a_index,
+                             a_exp_proc,
+                             ob,
+                             b,
+                             b_index + 1,
+                             b_exp_proc,
+                             status,
+                             exp,
+                             lem,
+                             cmp_infers);
```
Comment (ahmed-irfan) [inline src/theory/arith/nl_solver.cpp:503]: btw: I not happy with the clang-format here :(

Comment (ajreynol) [reply src/theory/arith/nl_solver.cpp:503]: Yeah, could refactor later so that there are less arguments here.

**Comment by ahmed-irfan** [comment]
the tests are failing

**Comment by ajreynol** [comment]
I believe this was due to build system issues which Mathias fixed.

**Pull request merged**